lua/entities/ent_jack_gmod_ezsolargenerator.lua

```diff
@@ -0,0 +1,247 @@
+        effectdata:SetScale(math.Rand(.5,1.5))
+        effectdata:SetRadius(math.Rand(2,4))
+        util.Effect("Sparks", effectdata)
+        self:EmitSound("items/suitchargeok1.wav", 80, 120)
+    end
+    
+    function ENT:ProducePower()
+        local SelfPos,Up,Forward,Right = self:GetPos(),self:GetUp(),self:GetForward(),self:GetRight()
+        local amt = math.min(math.floor(self:GetProgress()), self.MaxPower)
+
+        if amt <= 0 then return end
+        
+        local pos = SelfPos + Forward*15 - Up*25 - Right*2
+        for _, ent in pairs(ents.FindInSphere(pos, 100)) do
+            --print(ent, ent.GetResourceType and ent:GetResourceType())
+            if ((ent:GetClass() == "ent_jack_gmod_ezcrate") and (ent:GetResourceType() == "generic" 
+            or ent:GetResourceType() == "power") and (ent:GetResource() + amt <= ent.MaxResource)) then
+                    
+                if ent:GetResourceType() == "generic" then
+                    ent:ApplySupplyType("power")
+                end
+                    
+                ent:SetResource(math.min(ent:GetResource() + amt, ent.MaxResource))
+                self:SetProgress(self:GetProgress() - amt)
+                self:SpawnEffect(pos)
+                return
+            end
+        end
+        JMod.MachineSpawnResource(self, "power", amt, self:WorldToLocal(pos), Angle(-90, 0, 0), Up*-300)
+        self:SetProgress(self:GetProgress() - amt)
+        self:SpawnEffect(pos)
+    end
+
+    function ENT:CheckSky()
+        local HitAmount = 0
+        for i = 1, 10 do
+            for j = 1, 10 do
+                local StartPos = self:LocalToWorld(Vector(-5 + j*1, -100 + i*25, 10 + j*7.5))
+                local Dir = self:LocalToWorldAngles(Angle(260 - j*8, -10 + i*2, 0)):Forward()
+                local HitSky = util.TraceLine({start = StartPos, endpos = StartPos + Dir * 9e9, filter = {self}, mask = MASK_SOLID}).HitSky
+                if (HitSky) then HitAmount = HitAmount + 1 end
+                --JMod.Sploom(game.GetWorld(), StartPos + Dir * 1000, 0.5)
+            end
+        end
+        --print(HitAmount)
+        return HitAmount
+    end
+    
+    function ENT:TurnOn()
+        if (self:CheckSky() > 0) then
+            self:EmitSound("buttons/button1.wav", 60, 80)
+            self:SetState(STATE_ON)
+            self.NextUse = CurTime() + 1
+        else
+            self:EmitSound("buttons/button2.wav", 60, 100)
+        end
+    end
+
+    function ENT:TurnOff()
+        self:EmitSound("buttons/button18.wav", 60, 80)
+        self:SetState(STATE_OFF)
+        self:SetProgress(0)
+        self.NextUse = CurTime() + 1
+    end
+    
+    function ENT:Think()
+        local State = self:GetState()
+        if(State == STATE_ON)then
+            --stormfox support
+            if(StormFox)then 
+                if StormFox.GetWeather() ~= "Clear" or StormFox.IsNight() then self:NextThink(CurTime() + 30) return end
```

Comment:
as i said, we should just lower the charge rate rather than delay the think
just for math simplicity
the perf gains by doing this aren't worth the extra obfuscation

Reply:
Oh, I thought you said set it to 5 secs, or did i not push that? Shoot.

Reply:
Oh, I see what you mean, yeah, I got rid of that and didn't push it.